### PR TITLE
FE: Data Latency Graph

### LINKED ISSUE #1258: FE: Data Latency Graph
### Description

We want to have an additional graph to show the average data latency by log type.

More info on #1191

Mock TBD from #1257 
### Acceptance Criteria

- Implement a new graph or modify an existing one if possible
- Incorporate graph on Log Analysis Overview

### PR DESCRIPTION
**Pull request by alexmylonas**
## Description

This PR includes the work needed to display an extra on Log Analysis Overview Page.

It will be displayed in a tab form alongside Events By Log Type graph (check screenshots)

It's important to understand that the values displayed are the cumulative average time between `p_event_time` and `p_parse_time`. This means that a source is sending events with delay (like S3 Server Access which sends events with approx 1 hour delay) the result data will include that (i.e 3600 sec) 

Note:The values returned are displayed in seconds.

Mocks [here](https://app.abstract.com/projects/558a8de6-7134-4c8b-91c5-62074bb1279b/branches/39dcb3e2-1e72-49b2-afab-3a59c8459f66/commits/0232de261faf52a76684d9ca71584e71074ae5ef/files/c2bbc1a8-4778-4cd6-93a5-4630d8868201/layers/0274D9AF-E231-4B93-BAE6-18C06596802B?mode=build&selected=3879095143-268CB0CA-EF6A-495A-948B-4F85C8792780)
Closes #1258 

### Video:
Screen Recording 2020-09-08 at 5 22 01 PM

### Screenshots
Screenshot 2020-09-10 at 3 55 44 PM
Screenshot 2020-09-10 at 3 56 11 PM

### Checklist:

- [x] Make sure you give an high-level overview of the problem & the solution
- [x] Make sure the related issue is referenced
- [x] Make sure screenshots are added if the UI is changed in any way
- [x] Make sure all of the significant logic is covered by tests
- [x] Make sure your changes are accessible (addition of aria-attributes & correct HTML semantics)
- [x] Make sure you reference the related design document if it exists
- [x] Make sure the changes are responsive (appearing decently on devices)
- [x] ~~Make sure telemetry is added (analytics & errors)~~

### CODE REVIEW

File: api/graphql/schema.graphql [hunk 1]
```diff
@@ -384,11 +384,21 @@ type Series {
   values: [Int]
 }
 
+type FloatSeries {
+  label: String
+  values: [Float]
```
Comment (alexmylonas) [inline api/graphql/schema.graphql:389]: This is the first time we are using `Float`, are we ok with that?

File: web/src/components/BorderedTab/BorderTabDivider.tsx [hunk 2]
```diff
@@ -22,7 +22,6 @@ import { Box } from 'pouncejs';
 const BorderTabDivider: React.FC = () => {
   return (
     <Box
-      position="absolute"
```
Comment (alexmylonas) [inline web/src/components/BorderedTab/BorderTabDivider.tsx:25]: This was causing Tab Divider to take whole page, i didnt notice any change on the `ProfileSettings` where we are using this as well.

File: web/src/pages/LogAnalysisOverview/LogAnalysisOverview.tsx [hunk 3]
```diff
@@ -22,12 +22,12 @@ import withSEO from 'Hoc/withSEO';
 import TablePlaceholder from 'Components/TablePlaceholder';
 import { extractErrorMessage, getCurrentDate, subtractDays } from 'Helpers/utils';
 import Panel from 'Components/Panel';
-import EventsByLogType from 'Pages/LogAnalysisOverview/EventsByLogType';
 import AlertsTable from 'Pages/LogAnalysisOverview/AlertsTable';
 import LogAnalysisOverviewPageSkeleton from './Skeleton';
 import { useGetLogAnalysisMetrics } from './graphql/getLogAnalysisMetrics.generated';
 import AlertsBySeverity from './AlertsBySeverity';
 import AlertSummary from './AlertSummary';
+import SecondRowTabs from './SecondRowTabs';
```
Comment (alexmylonas) [inline web/src/pages/LogAnalysisOverview/LogAnalysisOverview.tsx:30]: Any suggestion for better naming this component?

**Comment by jacknagz** [comment]
@alexmylonas I don't think you need the word 'latency' after each log type in the legend

File: web/src/pages/LogAnalysisOverview/LogAnalysisOverview.tsx (hunk 3 above)
Comment (3nvi) [reply web/src/pages/LogAnalysisOverview/LogAnalysisOverview.tsx:30]: They are charts that have to do with log types, so `LogTypeCharts` or `LogTypesChartsSection` or something like that?

File: web/src/pages/LogAnalysisOverview/Skeleton/Skeleton.tsx [hunk 4]
```diff
@@ -33,7 +33,7 @@ const ComplianceOverviewPageSkeleton: React.FC = () => {
           </Panel>
         </SimpleGrid>
         <SimpleGrid columns={1} spacingX={3} spacingY={2} mb={3}>
-          <Panel title="Events by Log Type">
+          <Panel title="Events by Log Type | Data Latency">
```
Comment (3nvi) [inline web/src/pages/LogAnalysisOverview/Skeleton/Skeleton.tsx:36]: I think  we should show the during loading instead of changing the Pane's title. Can we add a "dummy" tabber here whose content is just `TablePlaceholder` components (along with the dark background of the charts)

File: web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+interface SecondRowTabsProps {
+  eventsProcessed: SeriesData;
+  eventsLatency: SeriesData;
+}
+
+const SecondRowTabs: React.FC<SecondRowTabsProps> = ({ eventsProcessed, eventsLatency }) => {
+  return (
+    <Card as="section" width={1}>
+      <Tabs width={1}>
```
Comment (3nvi) [inline web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx:34]: There is a mistake in the pounce typings.  Normally `Tabs` shouldn't accept any styling props (since they are not a styling component).

We should remove `width` from there and add it in a pounce style-related component

File: web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+        <Box>
+          <TabList>
+            <Tab>
+              {({ isSelected, isFocused }) => (
+                <BorderedTab isSelected={isSelected} isFocused={isFocused}>
+                  <Heading size="x-small" as="h4">
```
Comment (3nvi) [inline web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx:40]: BorderedTab should contain the styles inside of it. Can we make sure we add this styles as part of the `BorderedTab`  (`ProfileSettings` should have the same styles as well, although by mistake it doesn't 😆 )

File: web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+                  </Heading>
+                </BorderedTab>
+              )}
+            </Tab>
+          </TabList>
+          <BorderTabDivider />
+        </Box>
+        <Box p={6}>
+          <TabPanels>
+            <TabPanel unmountWhenInactive={true} lazy={true}>
```
Comment (3nvi) [inline web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx:60]: nitpick: change `unmountWhenInactive={true} lazy={true}` to `unmountWhenInactive lazy` 😄

File: web/src/pages/LogAnalysisOverview/EventsByLatency/EventsByLatency.tsx [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    }),
+    [timestamps, series]
+  );
+  return (
+    <Flex
+      data-testid="events-by-latency"
+      height="100%"
+      pt={4}
+      px={4}
+      backgroundColor="navyblue-500"
```
Comment (3nvi) [inline web/src/pages/LogAnalysisOverview/EventsByLatency/EventsByLatency.tsx:43]: Since down the line, you wrap this with a `<Box height={200}` can't we use the same element  to add any styles you want?

Right now we have a wrapping `Box` and this `Flex` here. Can't we "merge" those 2 together so that **this** component only renders a `<TimeSeriesChart />`?

File: web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+                    Events by Log Type
+                  </Heading>
+                </BorderedTab>
+              )}
+            </Tab>
+            <Tab>
+              {({ isSelected, isFocused }) => (
+                <BorderedTab isSelected={isSelected} isFocused={isFocused}>
+                  <Heading size="x-small" as="h4">
+                    Data Latency
```
Comment (3nvi) [inline web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx:50]: The mocks say `Data Latency by Log Type`. Just making sure that this wording is intentional and not a typo

File: web/src/components/BorderedTab/BorderedTab.tsx [hunk 10]
```diff
@@ -48,6 +48,7 @@ const BorderedTab: React.FC<BorderedTabProps> = ({ isSelected, isFocused, childr
     <Box
       mx={4}
       borderBottom="3px solid"
+      width={1}
```
[the remark this answers is not in the document]
Comment (3nvi) [reply web/src/components/BorderedTab/BorderedTab.tsx:51]: You can see that the outline is a bit "wrong" when you click the button (it stretches to the left 😛 )

File: web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+                </BorderedTab>
+              )}
+            </Tab>
+          </TabList>
+          <BorderTabDivider />
+        </Box>
+        <Box p={6}>
+          <TabPanels>
+            <TabPanel unmountWhenInactive={true} lazy={true}>
+              <Box height={200}>
```
Comment (3nvi) [inline web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx:61]: I think height is `289px` in designs

File: web/src/components/BorderedTab/BorderedTab.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (3nvi) [reply web/src/components/BorderedTab/BorderedTab.tsx:51]: You can also see that the "blue line" doesn't match the text width, but stretches out of it

**Review by 3nvi — COMMENTED** [review]
Looks great  @alexmylonas 🎉 

I have a few comments but nothing super strong that should block the release, please address and ping for a final re-review 😄

File: web/src/pages/LogAnalysisOverview/EventsByLatency/EventsByLatency.tsx (hunk 8 above)
Comment (vorillaz) [reply web/src/pages/LogAnalysisOverview/EventsByLatency/EventsByLatency.tsx:43]: Apologies for jumping into this but soon we will fix this one due to the upcoming chart enhancement PR 🤗

File: web/src/pages/LogAnalysisOverview/LogAnalysisOverview.test.tsx [hunk 12]
```diff
@@ -89,5 +90,11 @@ describe('Log Analysis Overview', () => {
 
     expect(alertsChart).toBeTruthy();
     expect(eventChart).toBeTruthy();
+
+    // Checking tab click works and renders Data Latency chart
+    const latencyChartTabButton = getByText('Data Latency');
+    fireEvent.click(latencyChartTabButton);
+    const latencyChart = getByTestId('events-by-latency');
```
Comment (vorillaz) [inline web/src/pages/LogAnalysisOverview/LogAnalysisOverview.test.tsx:97]: I guess assertions against `latencyChart` should be a bit more declarative.
Using `toBeTruthy` validates that the latency chart is not `false`, `undefined`, `null`, `Nan`, `''` or `0`.
You can always use the [`toBeInDocument`](https://github.com/testing-library/jest-dom#tobeinthedocument) or the [`toBeVisible`](https://github.com/testing-library/jest-dom#tobevisible) Jest matchers.

And IMHO it makes more sense to express some business logic using tests, you can always snapshot test the entire UI but a few selected checks can actually help the upcoming maintainers.

File: api/graphql/schema.graphql (hunk 1 above)
Comment (alexmylonas) [reply api/graphql/schema.graphql]: Good call out, added `!` where needed.

File: web/src/pages/LogAnalysisOverview/LogAnalysisOverview.tsx (hunk 3 above)
Comment (alexmylonas) [reply web/src/pages/LogAnalysisOverview/LogAnalysisOverview.tsx]: Renamed to `LogTypeCharts`

**Comment by alexmylonas** [comment]
> @alexmylonas I don't think you need the word 'latency' after each log type in the legend

All labels are coming from backend, @nhakmiller could you make a quick commit to remove the latency in the label field, or is something more complex?

File: api/graphql/schema.graphql [hunk 13]
```diff
@@ -384,11 +384,21 @@ type Series {
...
+}
+
 type SeriesData {
   timestamps: [AWSDateTime]
   series: [Series]
 }
 
+type FloatSeriesData {
```
[the remark this answers is not in the document]
Comment (alexmylonas) [reply api/graphql/schema.graphql:397]: Added

File: api/graphql/schema.graphql [hunk 14]
```diff
@@ -397,6 +407,7 @@ type SingleValue {
 type LogAnalysisMetricsResponse {
   eventsProcessed: SeriesData
   alertsBySeverity: SeriesData
+  eventsLatency: FloatSeriesData
```
[the remark this answers is not in the document]
Comment (alexmylonas) [reply api/graphql/schema.graphql:410]: All those seriesData in general are optional, at this point we are asking for everything in 1 API call but we could request a subset of the metrics.

File: web/src/components/BorderedTab/BorderedTab.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (alexmylonas) [reply web/src/components/BorderedTab/BorderedTab.tsx]: Removed this, added some padding to match mocks better

File: web/src/components/charts/TimeSeriesChart/TimeSeriesChart.tsx [hunk 15]
```diff
@@ -50,8 +50,23 @@ interface TimeSeriesLinesProps {
    * @default 3600 * 1000 * 24
    */
   maxZoomPeriod?: number;
+
+  /**
+   * This is parameter determines if we need to display the values with an appropriate suffix
+   */
+  units?: 'seconds';
```
[the remark this answers is not in the document]
Comment (alexmylonas) [reply web/src/components/charts/TimeSeriesChart/TimeSeriesChart.tsx]: Updated this per your suggestion, have a look and let me know!

File: web/src/pages/LogAnalysisOverview/Skeleton/Skeleton.tsx (hunk 4 above)
Comment (alexmylonas) [reply web/src/pages/LogAnalysisOverview/Skeleton/Skeleton.tsx]: Implemented this, works good.

File: web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx (hunk 5 above)
Comment (alexmylonas) [reply web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx]: Removed it

File: web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx (hunk 7 above)
Comment (alexmylonas) [reply web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx]: Done!

File: web/src/pages/LogAnalysisOverview/EventsByLatency/EventsByLatency.tsx (hunk 8 above)
Comment (alexmylonas) [reply web/src/pages/LogAnalysisOverview/EventsByLatency/EventsByLatency.tsx:43]: Thanks @vorillaz, i am keeping it like this for the moment then.

File: web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx (hunk 9 above)
Comment (alexmylonas) [reply web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx]: Updated, looks a bit wordy but matches mocks 😛

File: web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx (hunk 11 above)
Comment (alexmylonas) [reply web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx]: Idd, but mocks include both some text above and zoom bar below, I used my better judgement to pick a height for this.

Also height of chart is limited to `149px` in the mocks

File: web/src/pages/LogAnalysisOverview/LogAnalysisOverview.test.tsx (hunk 12 above)
Comment (alexmylonas) [reply web/src/pages/LogAnalysisOverview/LogAnalysisOverview.test.tsx:97]: Good call out, changed this to `toBeInDocument`!

File: web/src/components/BorderedTab/BorderTabDivider.tsx (hunk 2 above)
Comment (alexmylonas) [reply web/src/components/BorderedTab/BorderTabDivider.tsx]: Re-instated `position="absolute"` and add `position="relative"` to a parent Box element

File: web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx (hunk 6 above)
Comment (alexmylonas) [reply web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx]: Hmm, i dont get your suggestion here, what styles are you referring to? `size="x-small"`?

Comment (3nvi) [reply web/src/pages/LogAnalysisOverview/SecondRowTabs/SecondRowTabs.tsx]: `<Heading size="x-small" as="h4">` all of it. Let's put it as part of `BorderedTab`

File: api/graphql/schema.graphql (hunk 14 above)
[the remark this answers is not in the document]
Comment (3nvi) [reply api/graphql/schema.graphql:410]: absense of `!` means "the field can be `null`" and not that the field is optional.

Can the value `eventsLatency: null` be found?

**Review by 3nvi — APPROVED** [review]

**Comment by nhakmiller** [comment]
I don't have this deployed, but this should remove `latency` from the end of each label name. Feel free to give it a look.
